riscv/src/runtime.rs

```diff
@@ -148,6 +148,33 @@ impl Runtime {
         self.syscalls.contains_key(&s)
     }
 
+    pub fn with_keccak(mut self) -> Self {
+        self.add_submachine(
+            "std::machines::hash::keccakf::KeccakF",
+            None,
+            "keccakf",
+            [format!(
+                "instr keccakf {} ~ keccakf.keccakf {}, {}, STEP ->;", // instr keccakf X, Y -> ~ keccakf X, Y, STEP ->; (sub calls asm)
+                instr_register_params(0, 2, 0), // one pointer for input and one pointer for output (both as input arguments)
+                reg(0),
+                reg(1),
+            )],
+            0,
+            std::iter::once(format!("{} <=X= 0x100;", reg(0))) // filler value for input pointer
+                .chain(std::iter::once(format!("{} <=X= 0x300;", reg(1)))) // filler value for output pointer (at least 200 bytes away)
+                .chain(std::iter::once("keccakf;".to_string())) // must be called at least once
+                .chain((0..50).flat_map(|i| store_word(&reg(1), i as u32 * 4, "x0"))), // zero out 200 bytes following output pointer
```

Comment:
Can you explain why this is needed? Also where do the hard-coded values `0x100` and `0x300` come from?

Reply:
This is the `init` call, which is required as all submachine must be called at least once. Hardcoded values are arbitrary input and output address values as long as they are at least 200 bytes away from each other, because the input and output are both 200 bytes. These are set up so that we can call `keccakf`. Finally we need to zero out the output values.